File: src/lib/allocation.ts
import { config } from "./config";

export type RiskTolerance = "conservative" | "moderate" | "aggressive";

export interface AllocationInput {
  age: number;
  riskTolerance: RiskTolerance;
  investmentHorizonYears: number;
}

export interface AllocationResult {
  equityPct: number;
  debtPct: number;
  goldPct: number;
}

// No v1 source for the base equity formula or the gold sleeve (see
// decisions/log.md): base equity is the standard "constant minus age"
// glide path, and gold is a flat diversification sleeve capped by
// whatever equity leaves behind, so the three buckets always sum to 100.
export function computeAllocation({
  age,
  riskTolerance,
  investmentHorizonYears,
}: AllocationInput): AllocationResult {
  const baseEquityPct = config.base_equity_age_constant - age;
  let equityPct = baseEquityPct * config.risk_tolerance_multipliers[riskTolerance];

  if (investmentHorizonYears <= config.short_horizon_threshold) {
    equityPct -= config.short_horizon_shift;
  }

  equityPct = Math.min(100, Math.max(0, equityPct));

  const goldPct = Math.min(config.gold_allocation_pct, 100 - equityPct);
  const debtPct = 100 - equityPct - goldPct;

  return { equityPct, debtPct, goldPct };
}

export type AllocationBucket = "equity" | "debt" | "gold";

export type FundCategory =
  | "equity_large_cap"
  | "equity_diversified"
  | "debt_short_duration"
  | "fixed_deposit"
  | "gold_etf"
  | "sovereign_gold_bond";

export interface FundReferenceLike {
  id: string;
  category: FundCategory;
  aumCr: number;
}

// Section 5.2's example: equity_pct -> "large-cap index fund or diversified
// equity mutual fund category" implies each bucket spans 2 FundCategory
// enum values, not 1 - mirrored here explicitly since the plan never
// spells the mapping out as a table.
const BUCKET_CATEGORIES: Record<AllocationBucket, FundCategory[]> = {
  equity: ["equity_large_cap", "equity_diversified"],
  debt: ["debt_short_duration", "fixed_deposit"],
  gold: ["gold_etf", "sovereign_gold_bond"],
};

export function selectFundExamples<T extends FundReferenceLike>(
  bucket: AllocationBucket,
  funds: T[],
  count: number = config.fund_examples_per_category,
): T[] {
  const categories = BUCKET_CATEGORIES[bucket];
  return funds
    .filter((fund) => categories.includes(fund.category))
    .sort((a, b) => b.aumCr - a.aumCr)
    .slice(0, count);
}

File: backend/lib/allocation.py
from lib.finance_config import (
    BASE_EQUITY_AGE_CONSTANT, RISK_TOLERANCE_MULTIPLIERS,
    SHORT_HORIZON_THRESHOLD, SHORT_HORIZON_SHIFT, GOLD_ALLOCATION_PCT,
    FUND_EXAMPLES_PER_CATEGORY, GROWTH_SHARE_BASE_PCT, MID_SHARE_OF_GROWTH,
)

# Section 5.2's example: equity_pct -> "large-cap index fund or diversified
# equity mutual fund category" implies each bucket spans 2 category values,
# not 1 - mirrored here explicitly since the plan never spells the mapping
# out as a table.
BUCKET_CATEGORIES = {
    "equity": ["equity_large_cap", "equity_diversified"],
    "debt": ["debt_short_duration", "fixed_deposit"],
    "gold": ["gold_etf", "sovereign_gold_bond"],
}


# No v1 source for the base equity formula or the gold sleeve (see
# decisions/log.md): base equity is the standard "constant minus age"
# glide path, and gold is a flat diversification sleeve capped by
# whatever equity leaves behind, so the three buckets always sum to 100.
def compute_allocation(age: int, risk_tolerance: str, investment_horizon_years: int) -> dict:
    base_equity_pct = BASE_EQUITY_AGE_CONSTANT - age
    equity_pct = base_equity_pct * RISK_TOLERANCE_MULTIPLIERS[risk_tolerance]

    if investment_horizon_years <= SHORT_HORIZON_THRESHOLD:
        equity_pct -= SHORT_HORIZON_SHIFT

    equity_pct = min(100, max(0, equity_pct))

    gold_pct = min(GOLD_ALLOCATION_PCT, 100 - equity_pct)
    debt_pct = 100 - equity_pct - gold_pct

    return {"equity_pct": equity_pct, "debt_pct": debt_pct, "gold_pct": gold_pct}


def select_fund_examples(bucket: str, funds: list[dict], count: int | None = None) -> list[dict]:
    if count is None:
        count = FUND_EXAMPLES_PER_CATEGORY
    categories = BUCKET_CATEGORIES[bucket]
    matching = [f for f in funds if f["category"] in categories]
    matching.sort(key=lambda f: f["aum_cr"], reverse=True)
    return matching[:count]


def _equity_shares(risk_tolerance: str) -> tuple[float, float, float]:
    """Large / mid / small share of the equity slice, in percent. A higher risk tolerance moves weight from large into mid and small."""
    growth_share = min(100.0, GROWTH_SHARE_BASE_PCT * RISK_TOLERANCE_MULTIPLIERS[risk_tolerance])
    mid_share = growth_share * MID_SHARE_OF_GROWTH
    return 100.0 - growth_share, mid_share, growth_share - mid_share


# No source for these shares (see the spec, section 6.1): a rule of thumb keyed off the risk multiplier already in finance_config.
def compute_equity_split(risk_tolerance: str, equity_pct: float) -> dict:
    large_share, mid_share, _ = _equity_shares(risk_tolerance)
    large_pct = round(equity_pct * large_share / 100, 1)
    mid_pct = round(equity_pct * mid_share / 100, 1)
    # small takes the remainder so the three always add back up to the equity slice
    small_pct = round(equity_pct - large_pct - mid_pct, 1)
    return {"large_pct": large_pct, "mid_pct": mid_pct, "small_pct": small_pct}
